articles/flask-crud-api/index.md

```diff
@@ -0,0 +1,306 @@
+
+`db = SQLAlchemy(app)` creates an object of the SQLAlchemy and stores it in a variable `db`.
+
+### Creating the Todo model
+We are going to the `Todo` model which will be a Python class representing the `Todo` table in the database.
+
+```python
+# Model
+class Todo(db.Model):
+    __tablename__ = "todos"
+    id = db.Column(db.Integer, primary_key=True)
+    title = db.Column(db.String(20))
+    todo_description = db.Column(db.String(100))
+
+    def create(self):
+        db.session.add(self)
+        db.session.commit()
+        return self
+
+    def __init__(self, title, todo_description):
+        self.title = title
+        self.todo_description = todo_description
+
+    def __repr__(self):
+        return f"{self.id}"
+
+
+db.create_all()
+```
+`Todo` model class has an id property which is used as the primary key.</br> 
+`db.create_all()` makes the application to create all the defined tables in the database.</br> 
+`db.session.add(self)` adds the `Todo` instance into the SQLAlchemy database connection session.</br> 
+`db.session.commit()` executes all the database operations that are available in the session.
+
+```python
+class TodoSchema(ModelSchema):
+    class Meta(ModelSchema.Meta):
+        model = Todo
+        sqla_session = db.session
+
+    id = fields.Number(dump_only=True)
+    title = fields.String(required=True)
+    todo_description = fields.String(required=True)
+```
+The `TodoSchema` class above makes it possible to return JSON from the Python objects from the SQLAlchemy.
+
+### Creating the API endpoints
+
+**GET endpoint**
+```python
+@app.route('/api/v1/todo', methods=['GET'])
+def index():
+    get_todos = Todo.query.all()
+    todo_schema = TodoSchema(many=True)
+    todos = todo_schema.dump(get_todos)
+    return make_response(jsonify({"todos": todos}))
+```
+`@app.route('/api/v1/todo', methods=['GET'])` defines the URL to the endpoint and the HTTP method allowed for the endpoint.
+The above endpoint returns a list of `todos`.
+`get_todos = Todo.query.all()` queries all the `todos` from the database.
+`todo_schema = TodoSchema(many=True)` and `todos = todo_schema.dump(get_todos)` serializes the objects from SQLAlchemy.
+`return make_response(jsonify({"todos": todos}))` returns a list of `todos` as JSON.
+
+```python
+@app.route('/api/v1/todo/<id>', methods=['GET'])
+def get_todo_by_id(id):
+    get_todo = Todo.query.get(id)
+    todo_schema = TodoSchema()
+    todo = todo_schema.dump(get_todo)
+    return make_response(jsonify({"todo": todo}))
+```
+The above endpoint allows for a GET request passing in the id of the `Todo` in the URL path. It returns a single `Todo` with the specified id.
+
+**POST endpoint**
+```python
+@app.route('/api/v1/todo', methods=['POST'])
+def create_todo():
+    data = request.get_json()
+    todo_schema = TodoSchema()
+    todo = todo_schema.load(data)
```

Comment:
the line `todo = todo_schema.load(data)` threw an error: `marshmallow.exceptions.ValidationError: {'_schema': ['Invalid input type.']}` on a POST request.

Reply:
I think the issue is with the JSON you passing. Please check the keys.
Sample POST request JSON payload.
```JSON
{
    "title":"Go to School",
    "todo_description":"Complete assignments at School"
}
```